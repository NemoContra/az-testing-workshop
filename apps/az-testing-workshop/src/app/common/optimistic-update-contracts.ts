import { Contract } from '@az-testing-workshop/shared/util/api-models';

export const optimisticUpdateContracts = (
  contracts: Contract[] | undefined,
  updatedContract: Contract
) =>
  contracts?.map((contract) =>
    contract.id === updatedContract.id ? updatedContract : contract
  );
