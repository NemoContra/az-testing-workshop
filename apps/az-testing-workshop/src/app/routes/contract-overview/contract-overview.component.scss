.spinner-outlet {
  height: 20px;
  display: flex;
  justify-content: flex-end;
}

contract-table {
  display: block;
  margin: 36px;
}
