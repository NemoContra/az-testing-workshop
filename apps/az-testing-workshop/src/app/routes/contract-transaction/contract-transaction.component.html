<nx-link routerLink="/"><nx-icon name="chevron-left" />Zurück</nx-link>
@if(store.errorCode()) {
<nx-error>An error occurred</nx-error>
}
<div class="spinner-outlet">
  @if(store.loading()) {
  <nx-spinner size="medium" />
  }
</div>
@if(store.contract(); as contract) {
<div class="content">
  <contract-display [contract]="contract" />
  <div>
    <nx-formfield label="Transaktion wählen">
      <nx-dropdown
        [value]="store.transactionType()"
        (valueChange)="selectTransactionType($event)"
        placeholder="Bitte wählen"
      >
        @for(transaktionType of transactionTypes; track transaktionType) {
        <nx-dropdown-item [value]="transaktionType">{{
          transaktionType | transactionType
        }}</nx-dropdown-item>
        }
      </nx-dropdown>
    </nx-formfield>
  </div>
  <div>
    @switch (store.transactionType();) { @case ('AenderungNachname') {
    <form (submit)="submitAenderungNachname(nachname.value, contract)">
      <nx-formfield label="Nachname">
        <input
          nxInput
          type="text"
          #nachname
          [value]="contract.person.lastname"
          data-testid="nachname"
        />
      </nx-formfield>
      <button nxButton type="submit">Absenden</button>
    </form>
    } @case ('Kuendigung') {
    <form (submit)="submitKuendigung(contract)">
      <button [disabled]="!!contract.end" nxButton type="submit">
        Kündigen
      </button>
      @if(contract.end) {
      <nx-error class="cancel-error"
        >Der Vertrag wurde bereits zum
        {{ contract.end | date }} gekündigt</nx-error
      >
      }
    </form>
    } }
  </div>
</div>

}
