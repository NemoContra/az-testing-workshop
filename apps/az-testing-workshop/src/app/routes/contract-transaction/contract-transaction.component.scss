nx-link {
  margin: 16px;
  display: flex;
  place-items: center;
}

.spinner-outlet {
  height: 20px;
  display: flex;
  justify-content: flex-end;
}

.content {
  max-width: 1200px;
  margin: auto;
}

.cancel-error {
  margin-top: 16px;
}
