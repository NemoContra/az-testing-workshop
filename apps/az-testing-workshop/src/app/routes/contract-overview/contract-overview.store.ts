import {
  patchState,
  signalStore,
  withHooks,
  withMethods,
  withState,
} from '@ngrx/signals';
import { Contract } from '@az-testing-workshop/shared/util/api-models';
import { rxMethod } from '@ngrx/signals/rxjs-interop';
import { pipe, switchMap, tap } from 'rxjs';
import { inject } from '@angular/core';
import { HttpErrorResponse } from '@angular/common/http';
import { tapResponse } from '@ngrx/operators';
import { ContractService } from '../../services/contract.service';

export type ContractOverviewState = {
  contracts: Contract[] | undefined;
  errorCode: number | undefined;
  loading: boolean;
  query: string;
};

export const initialContractOverviewState: ContractOverviewState = {
  contracts: undefined,
  errorCode: undefined,
  loading: false,
  query: '',
};

export const ContractOverviewStore = signalStore(
  { providedIn: 'root' },
  withState(initialContractOverviewState),
  withMethods((store, contractService = inject(ContractService)) => ({
    setQuery: (query: string) => patchState(store, { query }),
    getContracts: rxMethod<string>(
      pipe(
        tap(() => patchState(store, { loading: true })),
        switchMap((query) =>
          contractService.getContracts(query).pipe(
            tapResponse({
              next: (contracts) => patchState(store, { contracts }),
              error: ({ status }: HttpErrorResponse) =>
                patchState(store, { errorCode: status }),
              finalize: () => patchState(store, { loading: false }),
            })
          )
        )
      )
    ),
  })),
  withHooks(({ getContracts, query }) => ({
    onInit: () => {
      getContracts(query);
    },
  }))
);

export type ContractOverviewStore = InstanceType<typeof ContractOverviewStore>;
