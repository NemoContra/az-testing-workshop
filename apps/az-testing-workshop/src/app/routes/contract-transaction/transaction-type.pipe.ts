import { Pipe, PipeTransform } from '@angular/core';
import { TransactionsType } from '../../common/transaction-type';

const transactionTypeMap = {
  AenderungNachname: 'Änderung des Nachnamen',
  Kuendigung: 'Vertragskündigung',
} as const satisfies Record<TransactionsType, string>;

export const transformTransactionType = <T extends TransactionsType>(
  transactionType: T
) => transactionTypeMap[transactionType];

@Pipe({
  name: 'transactionType',
  standalone: true,
})
export class TransactionTypePipe implements PipeTransform {
  transform = transformTransactionType;
}
