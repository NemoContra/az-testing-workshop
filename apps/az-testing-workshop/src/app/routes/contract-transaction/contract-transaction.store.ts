import { patchState, signalStore, withMethods, withState } from '@ngrx/signals';
import { Contract } from '@az-testing-workshop/shared/util/api-models';
import { rxMethod } from '@ngrx/signals/rxjs-interop';
import { pipe, switchMap, tap } from 'rxjs';
import { tapResponse } from '@ngrx/operators';
import { HttpErrorResponse } from '@angular/common/http';
import { inject } from '@angular/core';
import { TransactionsType } from '../../common/transaction-type';
import { ContractService } from '../../services/contract.service';
import { ContractOverviewStore } from '../contraxt-overview/contract-overview.store';

export type ContractTransactionState = {
  contract: Contract | undefined;
  errorCode: number | undefined;
  loading: boolean;
  transactionType: TransactionsType | undefined;
};

export const initialContractTransactionState: ContractTransactionState = {
  contract: undefined,
  errorCode: undefined,
  loading: false,
  transactionType: undefined,
};

const optimisticUpdateContracts = (
  contracts: Contract[] | undefined,
  updatedContract: Contract
) =>
  contracts?.map((contract) =>
    contract.id === updatedContract.id ? updatedContract : contract
  );

export const ContractTransactionStore = signalStore(
  withState(initialContractTransactionState),
  withMethods(
    (
      store,
      contractService = inject(ContractService),
      { contracts, ...contractOverviewStore } = inject(ContractOverviewStore)
    ) => ({
      selectTransaction: (transactionType: TransactionsType) =>
        patchState(store, { transactionType }),
      getContract: rxMethod<string>(
        pipe(
          tap(() => patchState(store, { loading: true })),
          switchMap((id: string) =>
            contractService.getContract(id).pipe(
              tapResponse({
                next: (contract) => patchState(store, { contract }),
                error: ({ status }: HttpErrorResponse) =>
                  patchState(store, { errorCode: status }),
                finalize: () => patchState(store, { loading: false }),
              })
            )
          )
        )
      ),
      updateContract: rxMethod<Contract>(
        pipe(
          tap(() => patchState(store, { loading: true })),
          switchMap((contract) =>
            contractService.updateContract(contract).pipe(
              tapResponse({
                next: (contract) => {
                  patchState(store, { contract });
                  patchState(contractOverviewStore, {
                    contracts: optimisticUpdateContracts(contracts(), contract),
                  });
                },
                error: ({ status }: HttpErrorResponse) =>
                  patchState(store, { errorCode: status }),
                finalize: () => {
                  patchState(store, { loading: false });
                },
              })
            )
          )
        )
      ),
    })
  )
);
