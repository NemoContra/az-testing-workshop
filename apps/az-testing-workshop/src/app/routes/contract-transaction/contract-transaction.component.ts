import {
  ChangeDetectionStrategy,
  Component,
  inject,
  Input,
} from '@angular/core';
import { ContractTransactionStore } from './contract-transaction.store';
import { NxDropdownModule } from '@aposin/ng-aquila/dropdown';
import { NxFormfieldModule } from '@aposin/ng-aquila/formfield';
import { TransactionTypePipe } from './transaction-type.pipe';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxLinkModule } from '@aposin/ng-aquila/link';
import { Router, RouterLink } from '@angular/router';
import { NxInputModule } from '@aposin/ng-aquila/input';
import { Contract } from '@az-testing-workshop/shared/util/api-models';
import { NxSpinnerModule } from '@aposin/ng-aquila/spinner';
import { NxButtonModule } from '@aposin/ng-aquila/button';
import { FormsModule } from '@angular/forms';
import {
  transactionQueryParam,
  TransactionsType,
  transactionTypes,
} from '../../common/transaction-type';
import { DatePipe } from '@angular/common';
import { NxMessageToastService } from '@aposin/ng-aquila/message';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [
    NxDropdownModule,
    NxFormfieldModule,
    TransactionTypePipe,
    NxIconModule,
    NxLinkModule,
    RouterLink,
    NxInputModule,
    NxSpinnerModule,
    NxButtonModule,
    FormsModule,
    DatePipe,
  ],
  providers: [ContractTransactionStore],
  selector: 'contract-transaction',
  standalone: true,
  styleUrl: './contract-transaction.component.scss',
  templateUrl: './contract-transaction.component.html',
})
export default class ContractTransactionComponent {
  @Input() set id(id: string) {
    this.store.getContract(id);
  }

  @Input() set transaktion(transaktion: TransactionsType | undefined) {
    if (transaktion) {
      this.store.selectTransaction(transaktion);
    }
  }

  store = inject(ContractTransactionStore);
  readonly transactionTypes = transactionTypes;

  private router = inject(Router);
  private messageToastService = inject(NxMessageToastService);

  async selectTransactionType(type: TransactionsType) {
    await this.router.navigate([], {
      queryParams: transactionQueryParam(type),
    });
  }

  submitAenderungNachname(
    lastname: string,
    { person, ...contract }: Contract
  ): void {
    this.store.updateContract({ ...contract, person: { ...person, lastname } });
    this.messageToastService.open(
      'Änderung des Nachname erfolgreich durchgeführt',
      {
        context: 'success',
        duration: 5000,
      }
    );
  }

  submitKuendigung(contract: Contract): void {
    this.store.updateContract({ ...contract, end: new Date().toDateString() });
    this.messageToastService.open('Kündigung erfolgreich durchgeführt', {
      context: 'success',
      duration: 5000,
    });
  }
}
