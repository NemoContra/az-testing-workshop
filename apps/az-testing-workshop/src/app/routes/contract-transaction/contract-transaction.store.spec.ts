import { createServiceFactory, SpectatorService } from '@ngneat/spectator/jest';
import { ContractTransactionStore } from './contract-transaction.store';
import { ContractService } from '../../services/contract.service';
import { AsyncSubject, of } from 'rxjs';
import { mockContracts } from '@az-testing-workshop/shared/util/mock-data';
import { Contract } from '@az-testing-workshop/shared/util/api-models';
import { HttpErrorResponse, HttpStatusCode } from '@angular/common/http';
import { optimisticUpdateContracts } from '../../common/optimistic-update-contracts';
import { ContractOverviewStore } from '../contract-overview/contract-overview.store';
import { patchState, signalState } from '@ngrx/signals';

const mockUpdatedContract: Contract = {
  ...mockContracts[0],
  person: {
    firstname: 'Berthold',
    lastname: 'Heisterkamp',
    dateOfBirth: '1961-05-16',
  },
};

const mockUpdatedContracts: Contract[] = [
  mockUpdatedContract,
  mockContracts[1],
];

jest.mock('@ngrx/signals', () => {
  const { patchState, ...module } = jest.requireActual('@ngrx/signals');
  return { ...module, patchState: jest.fn(patchState) };
});

jest.mock('../../common/optimistic-update-contracts', () => ({
  optimisticUpdateContracts: jest.fn(() => mockUpdatedContracts),
}));

describe('ContractTransactionStore', () => {
  let spectator: SpectatorService<
    InstanceType<typeof ContractTransactionStore>
  >;

  const createService = createServiceFactory({
    service: ContractTransactionStore,
    providers: [
      {
        provide: ContractService,
        useFactory: () => ({
          getContract: jest.fn().mockReturnValue(of(mockContracts[0])),
          getContracts: jest.fn().mockReturnValue(of(mockContracts)),
          updateContract: jest.fn().mockReturnValue(of(mockContracts[0])),
        }),
      },
      {
        provide: ContractOverviewStore,
        useFactory: () =>
          signalState({
            contracts: mockContracts,
          }),
      },
    ],
  });

  beforeEach(() => (spectator = createService()));

  afterEach(() => jest.clearAllMocks());

  it('should select transaction', () => {
    spectator.service.selectTransaction('AenderungNachname');
    expect(spectator.service.transactionType()).toEqual('AenderungNachname');
    spectator.service.selectTransaction('Kuendigung');
    expect(spectator.service.transactionType()).toEqual('Kuendigung');
  });

  it('should getContract successfully', () => {
    const contract$ = new AsyncSubject<Contract>();
    spectator.inject(ContractService).getContract.mockReturnValue(contract$);
    spectator.service.getContract('123456789');

    expect(spectator.service.loading()).toEqual(true);
    expect(spectator.service.errorCode()).toEqual(undefined);
    expect(spectator.service.contract()).toEqual(undefined);

    contract$.next(mockContracts[0]);
    contract$.complete();

    expect(spectator.service.loading()).toEqual(false);
    expect(spectator.service.errorCode()).toEqual(undefined);
    expect(spectator.service.contract()).toEqual(mockContracts[0]);
  });

  it('should getContract with an error', () => {
    const contract$ = new AsyncSubject<Contract>();
    spectator.inject(ContractService).getContract.mockReturnValue(contract$);
    spectator.service.getContract('123456789');

    expect(spectator.service.loading()).toEqual(true);
    expect(spectator.service.errorCode()).toEqual(undefined);
    expect(spectator.service.contract()).toEqual(undefined);

    contract$.error(
      new HttpErrorResponse({ status: HttpStatusCode.InternalServerError })
    );

    expect(spectator.service.loading()).toEqual(false);
    expect(spectator.service.errorCode()).toEqual(500);
    expect(spectator.service.contract()).toEqual(undefined);
  });

  it('should updateContract successfully', () => {
    const contract$ = new AsyncSubject<Contract>();
    const contractService = spectator.inject(ContractService);
    const contractOverviewStore = spectator.inject(ContractOverviewStore);

    contractService.updateContract.mockReturnValue(contract$);

    spectator.service.updateContract(mockUpdatedContract);

    expect(spectator.service.loading()).toEqual(true);
    expect(spectator.service.errorCode()).toEqual(undefined);
    expect(spectator.service.contract()).toEqual(undefined);

    contract$.next(mockUpdatedContract);
    contract$.complete();

    expect(optimisticUpdateContracts).toHaveBeenCalledTimes(1);
    expect(optimisticUpdateContracts).toHaveBeenCalledWith(
      mockContracts,
      mockUpdatedContract
    );

    expect(patchState).toMatchSnapshot();

    expect(spectator.service.loading()).toEqual(false);
    expect(spectator.service.errorCode()).toEqual(undefined);
    expect(spectator.service.contract()).toEqual(mockUpdatedContract);
    expect(contractOverviewStore.contracts()).toEqual(mockUpdatedContracts);
  });

  it('should updateContract with an error', () => {
    const contract$ = new AsyncSubject<Contract>();
    const contractService = spectator.inject(ContractService);
    const contractOverviewStore = spectator.inject(ContractOverviewStore);

    contractService.updateContract.mockReturnValue(contract$);

    spectator.service.updateContract(mockUpdatedContract);

    expect(spectator.service.loading()).toEqual(true);
    expect(spectator.service.errorCode()).toEqual(undefined);
    expect(spectator.service.contract()).toEqual(undefined);

    contract$.error(
      new HttpErrorResponse({ status: HttpStatusCode.InternalServerError })
    );

    expect(optimisticUpdateContracts).not.toHaveBeenCalled();

    expect(patchState).toHaveBeenCalledTimes(3);
    expect(patchState).toHaveBeenNthCalledWith(1, expect.anything(), {
      loading: true,
    });
    expect(patchState).toHaveBeenNthCalledWith(2, expect.anything(), {
      errorCode: 500,
    });

    expect(spectator.service.loading()).toEqual(false);
    expect(spectator.service.errorCode()).toEqual(500);
    expect(spectator.service.contract()).toEqual(undefined);
    expect(contractOverviewStore.contracts()).toEqual(mockContracts);
  });
});
