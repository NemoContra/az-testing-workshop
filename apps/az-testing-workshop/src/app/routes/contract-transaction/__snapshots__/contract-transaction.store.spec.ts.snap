// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractTransactionStore should updateContract successfully 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "contract": [Function],
        "errorCode": [Function],
        "loading": [Function],
        "transactionType": [Function],
        Symbol(STATE_SIGNAL): [Function],
      },
      {
        "loading": true,
      },
    ],
    [
      {
        "contract": [Function],
        "errorCode": [Function],
        "loading": [Function],
        "transactionType": [Function],
        Symbol(STATE_SIGNAL): [Function],
      },
      {
        "contract": {
          "contractNumber": "1/2345678/9",
          "id": "123456789",
          "person": {
            "dateOfBirth": "1961-05-16",
            "firstname": "Berthold",
            "lastname": "Heisterkamp",
          },
          "premium": 42.42,
          "start": "2024-01-01",
        },
      },
    ],
    [
      [Function],
      {
        "contracts": [
          {
            "contractNumber": "1/2345678/9",
            "id": "123456789",
            "person": {
              "dateOfBirth": "1961-05-16",
              "firstname": "Berthold",
              "lastname": "Heisterkamp",
            },
            "premium": 42.42,
            "start": "2024-01-01",
          },
          {
            "contractNumber": "1/2345678/8",
            "id": "123456788",
            "person": {
              "dateOfBirth": "1995-08-21",
              "firstname": "Bart",
              "lastname": "Simpson",
            },
            "premium": 99,
            "start": "2024-02-01",
          },
        ],
      },
    ],
    [
      {
        "contract": [Function],
        "errorCode": [Function],
        "loading": [Function],
        "transactionType": [Function],
        Symbol(STATE_SIGNAL): [Function],
      },
      {
        "loading": false,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
